proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+`base fees + prioritization fees`.
+
+## Detailed Design
+
+Application fees enable dapp developer to decide who to rebate and who to
+penalize through runtime features. In particular, programs needs to be able
+to penalize, even if the application is not CPI'd to.
+
+The checks on the application fees will be taken care of by Solana runtime.
+Total application fees paid will be included in the transaction through a
+special `PayApplicationFees` similar to the `ComputeBudget` program so that it
+is easier to calculate the total amount of fees that will be paid. In case of
+nonpayment or partial payment, the program is never executed. This instruction
+cannot be CPI'ed into.
+
+Application developers will be able to rebate these fees through a new runtime
+mechanism, but only if the transaction actually succeeds, this way CU heavy
+simulation of the program while write-locking the respective accounts will
+require full fee payment.
+
+### Payment
+
+When the cluster receives a new transaction, `PayApplicationFees` instruction is
+decoded to calculate the total fees required for the transaction message. Then
+we verify that the fee-payer has a minimum balance of:
+`per-transaction base fees` + `prioritization fees` +
+`maximum application fees to be paid`
+
+If the payer does not have enough balance, the transaction is not scheduled
+and fails with an error `InsufficientFunds`. Consider this case the same as if
+the payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+application fees. For all accounts with application fees, the fees paid should
+be greater than the required fees. In case of overpayment, the difference is
+stored in a variable and will be paid back to the payer in any case. In case the
+application fees are insufficiently paid or not paid then the transaction fails
+with an error `ApplicationFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The application fees minus rebates and minus overpayment are transferred to
+the respecitve accounts lamport balance, from where the owner can collect them.
+
+### Configuration
+
+Application fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+The ledger will need to store the amount of lamports required for every account.
+Hence changes should be constrained like any account write to the owner of the
+account. All programs will need to implement special instructions so that the
+authority of the accounts can sign to update application fees on the accounts.
+
+Setting the amount to `0` disables application fees. The maximum amount of
+application fees that can be set for an account will be limited to a constant
+amount of SOL (e.g. 20 SOL) so that accounts cannot become inaccessible through
```

Comment:
SIMD should state the max amount explictly, not just "maybe 20 SOL" :)
It'd be nice to defend whatever max amount is chosen as well

Reply:
These are the constrainst:
1. it should be a high enough number in order to enable very decisive punishments, like in the pyth use-case. Millions of dollars of value at risk depend on that account being readable, so for instance 1 SOL would be too little (11 SOL / block).
1. it should be a low enough number that most users can afford recovery in case they lock themselves out of their own account. configuration of fees is not visible to the scheduler and write locks an account, so they would need to have the write lock fee as balance in their wallet prior to execution.

20 feels right, that might not be true in 2 years from now, hence I thought, it's best to keep this as a constant that we can modify through a feature gate and adapt to economies of scale. Will add this to the SIMD ofc., but curious what you think.

Reply:
@mschneider We can have limit to 4 SOLs so that we can store application fees in u32 instead of u64.

Reply:
Why not u64. Ideally we have no limit. If the one true whale decides to blow the entire supply of SOL on a single account write, hallelujah!

Reply:
Fixed this. Now PRAW fees will be calculated on millilamports per CU Requested.